eggtimer.js

```diff
@@ -1,326 +1,681 @@
-    GITHUB.gitdata.deleteReference(params, callback);
+function deleteReference(params) {
+    return new Promise( (resolve, reject) => {
+        Github.authenticate(GithubAuthentication);
+        Github.gitdata.deleteReference(params, (err) => {
+            if (err) {
+                reject("Error! Could not delete reference: " + err);
+                return;
+            }
+            console.log("Deleted reference to sha:", params.ref);
+            resolve(true);
+       });
+    });
 }
 
-// Finds the PR URL associated with the given head SHA
-function lookupPullRequest(owner, repo, sha, callback) {
-    const params = {
-        owner: owner,
-        repo: repo
-    };
-    GITHUB.pullRequests.getAll(params, function(err, res) {
+function updatePR(params) {
+   return new Promise( (resolve, reject) => {
+     Github.authenticate(GithubAuthentication);
+     Github.pullRequests.update(params, (err, res) => {
         if (err) {
-            console.log("err with pr.get: " + err);
-            callback(err, null);
+            reject("Error! Could not update PR: " + err);
             return;
         }
+        console.log("Updated PR" + res.data.number, res.data.state);
+        resolve(res.data.state);
+     });
+  });
+}
 
-        for (var i in res.data) {
-            const pr = res.data[i];
-            if (pr.head.sha == sha) {
-                const url = pr.url;
-                callback(null, url);
-                return;
-            }
+function addLabels(params) {
+   return new Promise( (resolve, reject) => {
+     Github.authenticate(GithubAuthentication);
+     Github.issues.addLabels(params, (err, res) => {
+        if (err) {
+            reject("Error! Could not add label to PR" + params.number + ": " + err);
+            return;
         }
-
-        callback("PR not found: (" + owner + ", " + repo + ", " + sha + ")", null);
-    });
+        console.log("PR" + params.number, "labels total:", res.data.length);
+        for (let label of res.data)
+           console.log("PR label name:", label.name);
+        resolve(true);
+     });
+  });
 }
 
-function parsePullRequestUrl(url) {
-    const re = /^https?:\/\/([^\/]+)\/repos\/([^\/]+)\/([^\/]+)\/pulls\/(\d+)$/;
-    const match = re.exec(url);
-    return {
-        owner: match[2],
-        repo: match[3],
-        number: match[4]
-    };
+function removeLabel(params) {
+   return new Promise( (resolve) => {
+     Github.authenticate(GithubAuthentication);
+     Github.issues.removeLabel(params, (err) => {
+        if (err)
+            console.log("Could not remove label " + params.name + " to PR" + params.number + ": " + err);
+        resolve(true);
+     });
+  });
 }
 
-function isMergeable(obj) {
-    return 'mergeable' in obj && !!obj.mergeable;
+function finishMerging(prContext) {
+    let statusParams = commonParams();
+    assert(prContext.autoSha);
+    statusParams.ref = prContext.autoSha;
+    let errorLabel = MergeFailedLabel;
```

Comment:
```
If MergeFailedLabel should be set at all (discussed elsewhere), then it should be set when we fail to either declare the PR not ready for merging or merge it.
```

In my understanding, we should set this label in PR if and only if the last stage (finishMerging()) is OK, actually meaning that auto tests have been failed.
However, I probably I am misunderstanding this question. Note, that in bfc2af0 I have changed that code, adding a simple flag to distinguish two situations in finishMerging():
* Reject because not all auto_checks have been completed (without PR labeling)
* Reject on an error or failed auto_checks (with PR labeling).

Reply:
Let's revisit this after the next rewrite. Just keep in mind that 

* we want humans to be able to see important PR states at any given time _but_ we do not want to bombard humans with temporary, unimportant, and/or duplicated information. 
* the bot should not get into a livelock situation where the bot repeats the same (failing) action over and over again.

From those principles point of view, a `MergeFailedLabel` that only reflects the "last stage is OK, [but] auto tests have failed" state is probably not needed because the state of auto tests is already visible on human-oriented GitHub pages and is available to the bot via GitHub API -- no special label is needed to reflect all that.

My "fail to either declare the PR not ready for merging or merge it" conditions are more general and do not contradict the above principles. In the more general context, "merging" should be interpreted as "advancing" (i.e., making progress towards merging or actually merging).